### PR TITLE
Fake PR for reviewing purposes

### PR DESCRIPTION
**Pull request by lsago**
This is just so it's easier to comment on different lines, etc. Not meant to be merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,376 @@
...
+The library can be included in your Vantage6 algorithm by listing it in the `requirements.txt` and `setup.py` file of
+your
+algorithm.
+
+## Including the library in your Vantage6 algorithm
+
+For the `requirements.txt` file, you can add the following line to the file:
+
+```
+git+https://github.com/STRONGAYA/v6-tools-general.git@v1.0.2
```
Comment (bartvanb) [inline README.md]: why specify the version here? Are you going to update this every time a new version becomes available?

Comment (bartvanb) [reply README.md]: same below in some places

File: README.md [hunk 2]
```diff
@@ -1 +1,376 @@
...
+
+Some empirical tests may occasionally fail due to the inherent mathematical differences between federated and
+centralised computations:
+
+**Empirical Tests (tests/empirical/)**:
+
+- `test_single_organisation_equivalence`: May fail due to division-by-zero issues in quantile calculations for single
+  organisations. This doesn't affect multi-organisation federated scenarios.
+- `test_mixed_distribution_quantiles`: May fail when organisations have very different data distributions, as federated
+  quantiles mathematically differ from centralised ones when internal distributions vary significantly.
```
Comment (bartvanb) [inline README.md:334]: Are these unit tests or tests to be run on data outside of this repo? If they are on test data generated within the repo, is it not easier to just create a dummy data set in the same way for every test round?

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+requires-python = ">=3.10"
+license = "Apache-2.0"
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "Operating System :: OS Independent",
+]
+dependencies = [
+    "numpy>=2.1.2",
+    "pandas>=2.2.3",
+    "vantage6-algorithm-tools>=3.6.0",
```
Comment (bartvanb) [inline pyproject.toml]: maybe require at least v4? this minimal version is very old

File: src/vantage6_strongaya_general/__pycache__/__init__.cpython-312.pyc
Comment (bartvanb) [inline src/vantage6_strongaya_general/__pycache__/__init__.cpython-312.pyc:1]: can you add `*.pyc` to .gitignore and then untrack these files?

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+
+# Type definitions for variable details (data type setting)
+class CategoricalDetails(TypedDict):
+    datatype: str
+    inliers: List[str]
+
+
+class NonCategoricalDetails(TypedDict):
+    datatype: str
+    inliers: List[int | float]
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:66]: maybe add docstrings to these classes? It is now not clear to me what `inlier` is

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+        organisations = client.organization.list()
+        return [organisation.get("id") for organisation in organisations]
+
+    if isinstance(organisation_ids, list):
+        try:
+            # Attempt to convert all elements to integers
+            organisation_ids = [int(i) for i in organisation_ids]
+        except ValueError:
+            safe_log("error", "Organisation IDs should be a list of integers")
+            return []
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py]: Why return an empty list here? Won't that lead to issues later when the central function tries to create subtasks for 0 orgs and then gets an error? Maybe you can raise a `UserInputError` here - see https://github.com/vantage6/vantage6/blob/main/vantage6-algorithm-tools/vantage6/algorithm/tools/exceptions.py

Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py]: Same question for line 181

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+    """
+    # Only allow specific variable names to be logged, not values
+    if variables:
+        # If more than 5 variables, truncate the list to prevent data mining
+        if len(variables) > 5:
+            var_str = ", ".join(variables[:5]) + f" and {len(variables) - 5} more"
+        else:
+            var_str = ", ".join(variables)
+
+        message = message.replace("{variables}", var_str)
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:204]: just curious: where can I find an example where `variables` is specified? That seems to be the main 'safe' part of this fucntion, curious what you use it for. 

This may also be transferred to the general vantage6 algorithm tools if useful and generalizable

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+
+    Returns:
+        T: Result of calculation_func or default_value if it fails
+    """
+    try:
+        return calculation_func(**kwargs)
+    except Exception as e:
+        # Log generic error without data details
+        safe_log("warn", f"Calculation error: {type(e).__name__}. Using default value")
+        return cast(T, default_value)
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:242]: Also wondering here if you have an example where you're using it? The default could be quite tricky - would you return 0 or so? How do you track that an error is raised?

Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:242]: Maybe interesting to mention: in v5 we will provide a decorator `@handle_pandas_errors` to do a similar thing - https://github.com/vantage6/vantage6/blob/release/5.0/vantage6-algorithm-tools/vantage6/algorithm/tools/error_handling.py#L34. Feel free to comment on that to improve it so it is more useful for you

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+                    safe_log(
+                        "warn",
+                        f"Unexpected values detected in categorical variable '{variable}',"
+                        f"coercing outliers to missing values",
+                    )
+            else:
+                raise UserInputError(
+                    f"Unknown datatype '{datatype}' for variable '{variable}'"
+                )
+    return df
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:300]: general note for this function: maybe add error handling in case type casting fails?

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+        invalid_rows = (len(variables) - missing_count) < required_items
+        if invalid_rows.any():
+            error_msg = (
+                f"Insufficient valid responses for {new_variable_name}. "
+                f"Required: {required_items}, Found invalid rows: {df.index[invalid_rows].tolist()}"
+            )
+            if strict:
+                raise DataError(error_msg)
+            else:
+                safe_log("warn", error_msg)
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:356]: code would be easier to read if there were separate functions to check ranges and required items. 

I also see value in adding those kinds of functions to v6 algorithm tools package - maybe next step?

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 10]
```diff
@@ -0,0 +1,665 @@
...
+
+    Returns:
+        bool: True if all variables are present, False otherwise.
+    """
+    missing_vars = [var for var in variables if var not in df.columns]
+    if missing_vars:
+        safe_log(
+            "error",
+            f"Missing variables in DataFrame: {', '.join(missing_vars)}",
+        )
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:381]: I guess here you would want to supply the third arg to `safe_log` so that not all vars are printed?

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 11]
```diff
@@ -0,0 +1,665 @@
...
+        raise UserInputError(
+            "Not all specified variables were found in the working copy of the DataFrame, "
+            "ensure that spelling of input parameters is correct."
+        )
+    return True
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:386]: Would be nice to have this in something like `vantage6.algorithm.tools.data_checks`. Let's discuss if/how to create PR(s) to do that

Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:386]: Same for functions several functions below

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+        return True
+    else:
+        raise CollectResultsError(
+            "Unexpected results format received from the subtask. "
+            "Please check the algorithm input or the nodes' logs for more information."
+        )
+
+
+@pd.api.extensions.register_dataframe_accessor("predetermined_info")
+class PredeterminedInfoAccessor:
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:463]: Maybe put this class in its own file?

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+        update_with_output_index: Optional[int] = None,
+        **kwargs: Any,
+    ) -> None:
+        """
+        Add a custom statistic to the DataFrame.
+
+        Args:
+            stat_name (str): Name of the statistic to store
+            calculation_func (Optional[Callable[..., Any]]): Function to calculate the statistic
+            value (Any, optional): Direct value to store if not using calculation_func
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:514]: The number of parameters makes it a bit hard to follow what happens. My preference would be to split it into several functions with clear names, e.g.
```python
def calc_stat(stat_name, calculation_fnc):
    values = calc_func()
    self._add_stat(stat_name, values)

def add_precalculated_stat(stat_name, values):
    self._add_stat(stat_name, values)
```
etc

This makes the function itself also easier to read - now there is a long if statement where it's tricky to see where one usage scenario ends and another begins

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+                        if isinstance(result, tuple):
+                            stat_value = (
+                                result[store_output_index]
+                                if store_output_index is not None
+                                else result[0]
+                            )
+                            if update_with_output_index is not None:
+                                updated_columns[col] = result[update_with_output_index]
+                            return stat_value
+                        return result
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py:546]: Please don't define the subfunc within a for loop (I guess Python will redefine it every time?!). I think it's good practice to only use subfunctions when there is a clear goal, here they could also be private class functions

File: src/vantage6_strongaya_general/miscellaneous.py [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+
+        Args:
+            stat_name (str): Name of the statistic to retrieve
+            column (Optional[str]): If the stat is per-column, specify which column to get
+
+        Returns:
+            Any: The stored statistic value
+
+        Raises:
+            KeyError: If statistic not found
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/miscellaneous.py]: InputError is raised ;-)

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 12 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:463]: Another question: where have you been using this?

Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:463]: Also learned something new here, I didn't know pandas offerred this. Not entirely sure how I feel about storing data inside pandas rather than keeping it separate, but pretty cool feature!

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 16]
```diff
@@ -0,0 +1,1427 @@
+"""
+------------------------------------------------------------------------------
+Privacy Protection Utilities
+
+File organisation:
+- General data preprocessing utilities (e.g., masking variables, sample size thresholding)
+- Privacy budget management
+- Differential privacy mechanisms (Laplace and Gaussian)
+- Differential privacy application functions
+- Main DP function
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:10]: Since this file is pretty long, it would be an improvement to separate these functionalities in their own files.

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 17]
```diff
@@ -0,0 +1,1427 @@
...
+    Returns:
+       pd.DataFrame: The processed DataFrame with unnecessary variables removed.
+    """
+    safe_log("info", "Masking unnecessary variables in the working draft of DataFrame")
+
+    # Drop any irrelevant variables
+    for variable in df.columns:
+        if variable not in variables_to_use:
+            df = df.drop(columns=variable)
+            continue
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:63]: this continue is not doing anything

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 18]
```diff
@@ -0,0 +1,1427 @@
...
+    return df
+
+
+class PrivacyBudgetManager:
+    """
+    Manages the privacy budget to ensure overall privacy guarantees are maintained.
+
+    Tracks consumed privacy budget (epsilon) across multiple queries or operations and
+    prevents exceeding the total allocated budget.
+    """
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:123]: Two comments:
- A reference may be nice - I don't know this method and the docstring doesn't enable me to learn about it elsewhere
- The class would be more self-contained if it computed the epsilon itself, rather than just accepting a value that some other function gives it. That way it's not really a 'manager' but more a 'tracker'. Maybe some refactoring can improve this?

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 19]
```diff
@@ -0,0 +1,1427 @@
...
+        delta (float): The failure probability parameter.
+        epsilon (float): The privacy parameter (smaller means more privacy).
+
+    Returns:
+        float: The scale (standard deviation) for the Gaussian noise.
+    """
+    return sensitivity * math.sqrt(2 * math.log(1.25 / delta)) / epsilon
+
+
+def _generate_query_seed(
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:226]: I have a feeling there is a refactor possible here somewhere into a `class`. A class would probably have all the arguments provided here as attributes, so then it would be simply `self.generate_query_seed()` without needing to pass params around. Especially since I see much of the same params being passed around in several functions around here

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 20]
```diff
@@ -0,0 +1,1427 @@
...
+
+    Returns:
+        float: A noise value from the appropriate distribution.
+    """
+    # Create a deterministic RNG if seed is provided
+    if seed is not None:
+        # Create a separate random number generator that won't affect other code
+        rng = np.random.RandomState(seed)
+    else:
+        # Use the global RNG if no seed (less secure but maintains backwards compatibility)
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:311]: backwards compatible with what?

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 21]
```diff
@@ -0,0 +1,1427 @@
...
+        rng = np.random.RandomState()
+
+    if distribution == "laplace":
+        scale = _get_laplace_scale(sensitivity, epsilon)
+        return rng.laplace(0, scale)
+    else:  # gaussian
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:317]: consider throwing error if `distribution` is something else but laplace or gaussian

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 22]
```diff
@@ -0,0 +1,1427 @@
...
+    # Other potentially sensitive fields to remove
+    sensitive_fields = ["bin_edges", "categories"]
+    for field in sensitive_fields:
+        if field in safe_metadata:
+            # Instead of values, just store the count
+            if isinstance(safe_metadata[field], list):
+                safe_metadata[f"{field}_count"] = len(safe_metadata[field])
+            del safe_metadata[field]
+
+    return safe_metadata
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:359]: To make this more easily maintainable (and more easily readable), consider creating dataclasses and/or enums to make it explicit what the metadata structure is and which variables are deemed sensitive (and thus modified by this function)

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 23]
```diff
@@ -0,0 +1,1427 @@
...
+            return max_val
+        elif operation == "count":
+            # For count, sensitivity is 1
+            return 1.0
+        else:
+            # Default conservative approach
+            return value_range
+    else:
+        # For non-numeric data, use default sensitivity of 1
+        return 1.0
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:413]: you can lose an indent here to decrease function complexity by reversing check order:
```
if non numeric:
    return 1

# rest of your code that no longer needs to be inside an if
```

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 24]
```diff
@@ -0,0 +1,1427 @@
...
+
+    # Store security notice
+    dp_df.attrs["dp_metadata"]["security_notice"] = dp_results.get(
+        "security_notice", "This DataFrame contains privacy-protected data."
+    )
+
+    return dp_df
+
+
+def apply_differential_privacy(
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:1050]: I found my class (see comment above). 
```
dp_mgr = DifferentialPrivacyManager(...)
safe_df = db_mgr.apply_dp(df)
```
Something like this?
Probably saves you 100+ lines of passing parameters

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 25]
```diff
@@ -0,0 +1,1427 @@
...
+    else:
+        variables_list = variables
+
+    safe_log(
+        "info", f"Applying differential privacy to {len(variables_list)} variables"
+    )
+
+    # Create a budget manager if none was provided
+    if budget_manager is None:
+        budget_manager = PrivacyBudgetManager(epsilon, delta)
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:1165]: isn't it better to always reinitialize this? What if someone passes one that's initialized with different epsilon/delta?

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 26]
```diff
@@ -0,0 +1,1427 @@
...
+                if "metadata" in result:
+                    results[var_name]["metadata"] = _sanitise_metadata(
+                        result["metadata"]
+                    )
+
+    # Return based on specified return_type
+    if return_type == "dataframe":
+        return _get_dp_dataframe(df, results)
+    else:
+        return results
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:1427]: this function is quite long and could be refactored so that it is easier which steps are done

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 27]
```diff
@@ -0,0 +1,1427 @@
...
+            "sensitivity": sensitivity,
+            "noise_scale": noise_scale,
+            "distribution": distribution,
+            "bin_count": len(bins) - 1,
+            "bin_edges": (
+                bin_edges
+                if debug_mode
+                else [f"bin_edge_{i}" for i in range(len(bin_edges))]
+            ),
+        },
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:783]: I feel these results could be a dataclass?

File: src/vantage6_strongaya_general/privacy_measures.py [hunk 28]
```diff
@@ -0,0 +1,1427 @@
...
+                "security_warning",
+            ]:
+                if "metadata" in result:
+                    results[var_name]["metadata"] = _sanitise_metadata(
+                        result["metadata"]
+                    )
+
+    # Return based on specified return_type
+    if return_type == "dataframe":
+        return _get_dp_dataframe(df, results)
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/privacy_measures.py:1425]: How are you going to ensure this function gets updated properly when you update the fields in the results?

In my experience, that is difficult. It might be simpler to just choose one of the return values and stick with that

File: src/vantage6_strongaya_general/general_statistics.py [hunk 29]
```diff
@@ -0,0 +1,1415 @@
+"""
+------------------------------------------------------------------------------
+Statistical Analysis Functions
+
+File organisation:
+- Public API functions for computation
+- Orchestration functions (_orchestrate_*)
+- Basic statistical computation functions (_compute_local_*)
+- Aggregate computation functions (_compute_aggregate_*)
+------------------------------------------------------------------------------
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/general_statistics.py:10]: maybe create a directory statistics/ with files `compute_aggregate`, `orchestrate` etc?

File: src/vantage6_strongaya_general/general_statistics.py [hunk 30]
```diff
@@ -0,0 +1,1415 @@
...
+        adjusted_deviation = pd.DataFrame(columns=["variable", "statistic", "value"])
+
+        safe_log(
+            "warn",
+            "No variables to analyse for adjusted deviation due to lacking aggregate numerical statistics",
+        )
+    else:
+        adjusted_deviation = safe_calculate(
+            _orchestrate_local_adjusted_deviation,
+            pd.DataFrame(columns=["variable", "statistic", "value"]),
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/general_statistics.py:259]: You often use these strings throughout these files, but what if you every had to change them? You could use an enum to ensure you'd only need to change the name in one place.

File: src/vantage6_strongaya_general/general_statistics.py (hunk 29 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/general_statistics.py:10]: Or maybe put add some ordering to make it clear which central functions go with which federated funcs?

File: src/vantage6_strongaya_general/general_statistics.py [hunk 31]
```diff
@@ -0,0 +1,1415 @@
...
+    ]
+    numerical_columns = [
+        col for col in df.columns if pd.api.types.is_numeric_dtype(df[col])
+    ]
+
+    # Process categorical variables if any exist
+    if categorical_columns:
+        categorical_statistics = safe_calculate(
+            _orchestrate_local_categorical_statistics,
+            pd.DataFrame(columns=["variable", "value", "count"]),
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/general_statistics.py:202]: initializing a dataframe here looks odd to me

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 7 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:242]: Also, does this work if default_value is None?

File: src/vantage6_strongaya_general/general_statistics.py [hunk 32]
```diff
@@ -0,0 +1,1415 @@
...
+    Returns:
+        Union[int, float]: The local mean.
+    """
+    sum_val = safe_calculate(_compute_local_sum, 0.0, column_values=column_values)
+    rows = safe_calculate(
+        _compute_local_number_of_rows, 1, column_values=column_values, drop_na=True
+    )
+
+    # Avoid division by zero
+    return sum_val / rows if rows > 0 else 0.0
```
Comment (bartvanb) [inline src/vantage6_strongaya_general/general_statistics.py:894]: If there are too few rows to share data, where are you preventing this data from sharing?

File: src/vantage6_strongaya_general/general_statistics.py
Comment (bartvanb) [inline src/vantage6_strongaya_general/general_statistics.py:1]: My main worry for this file is: how is it checked that the output is OK to share? There seem to be no checks on the size of the data before using these functions. Is that meant to be handled by DP?

File: tests/TESTING_GUIDE.md [hunk 33]
```diff
@@ -0,0 +1,418 @@
...
+    result = mask_unnecessary_variables(test_data, variables_to_keep)
+    
+    # Assert
+    assert list(result.columns) == variables_to_keep
+    pd.testing.assert_series_equal(result['keep_me'], test_data['keep_me'])
+```
+
+### 2. Use Fixtures for Common Test Data
+
+```python
```
Comment (bartvanb) [inline tests/TESTING_GUIDE.md]: Why have you added all these unit tests here in the markdown? It seems to me that this will be impossible to maintain: whenever you update a unit test, you'll have to update it here as well, and in my experience almost everyone forgets that (at least sometimes).

For me the part at the end is most important as a reader: how can I run these tests? And maybe what is tests are implemented can be written down briefly (but this is subject to change and risks being outdated).

File: tests/TESTING_GUIDE.md [hunk 34]
```diff
@@ -0,0 +1,418 @@
...
+1. **Test Behavior, Not Implementation**: Focus on what the function should do, not how it does it
+2. **Use Descriptive Names**: Test names should explain the scenario being tested
+3. **One Assertion Per Concept**: Keep tests focused and easy to debug
+4. **Test Edge Cases**: Empty inputs, boundary values, error conditions
+5. **Use Fixtures**: Share common test data and setup code
+6. **Mock External Dependencies**: Keep tests fast and isolated
+7. **Validate Both Structure and Values**: Check return types and actual results
+8. **Test Error Conditions**: Ensure functions fail gracefully
+9. **Document Complex Tests**: Explain non-obvious test logic
+10. **Keep Tests Fast**: Unit tests should run in milliseconds
```
Comment (bartvanb) [inline tests/TESTING_GUIDE.md]: Who is your audience here? These are common principles, I prefer that they are left out.

**Review by bartvanb — COMMENTED** [review]
I have not yet gone over everything, but here some first comments. Overall nice work! I think some things could be implemented directly in vantage6, let's discuss that later.

File: pyproject.toml (hunk 3 above)
Comment (bartvanb) [reply pyproject.toml]: important. Also, specify <5 to prevent using that when it's out

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 5 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py]: important

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 7 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:242]: important

File: src/vantage6_strongaya_general/privacy_measures.py (hunk 25 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/privacy_measures.py:1165]: important

File: src/vantage6_strongaya_general/general_statistics.py (hunk 32 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/general_statistics.py:894]: important

File: src/vantage6_strongaya_general/general_statistics.py
Comment (bartvanb) [reply src/vantage6_strongaya_general/general_statistics.py:1]: important

File: tests/TESTING_GUIDE.md (hunk 33 above)
Comment (bartvanb) [reply tests/TESTING_GUIDE.md]: important

File: pyproject.toml (hunk 3 above)
Comment (JHogenboom) [reply pyproject.toml]: Resolved in 4f9e7fc94f1c95a540323bb6883b5cf71e894963

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 7 above)
Comment (JHogenboom) [reply src/vantage6_strongaya_general/miscellaneous.py:242]: There is an 'example' in the unit tests, which actually do not use a default value and work: https://github.com/STRONGAYA/v6-tools-general/blob/main/tests/unit/test_miscellaneous.py#L97-L146 

Should we put an example in the docstring of the function?

File: src/vantage6_strongaya_general/general_statistics.py (hunk 32 above)
Comment (JHogenboom) [reply src/vantage6_strongaya_general/general_statistics.py:894]: This is not handled in this function - as you noticed

The idea was to provide standalone functions that solely 'do the math' which are then for example preceeded or followed by privacy measures in the partial; where such cases should be adequately testing in some kind of automated tests e.g. here: https://github.com/STRONGAYA/v6-descriptive-statistics/blob/main/tests/integration/test_algorithm_integration.py#L166-L196

File: src/vantage6_strongaya_general/general_statistics.py
Comment (JHogenboom) [reply src/vantage6_strongaya_general/general_statistics.py:1]: Same as [above](https://github.com/STRONGAYA/v6-tools-general/pull/8#discussion_r3443420803)
> This is not handled in this function - as you noticed
> 
> The idea was to provide standalone functions that solely 'do the math' which are then for example preceeded or followed by privacy measures in the partial; where such cases should be adequately testing in some kind of automated tests e.g. here: https://github.com/STRONGAYA/v6-descriptive-statistics/blob/main/tests/integration/test_algorithm_integration.py#L166-L196

File: src/vantage6_strongaya_general/__pycache__/__init__.cpython-312.pyc
Comment (JHogenboom) [reply src/vantage6_strongaya_general/__pycache__/__init__.cpython-312.pyc:1]: Resolved in 336cca95657fe41f4536061fc548838111b67662

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 5 above)
Comment (JHogenboom) [reply src/vantage6_strongaya_general/miscellaneous.py]: Resolved in 03ceee493a67a326858f0afc6835dace6b631c33

File: tests/TESTING_GUIDE.md (hunk 33 above)
Comment (JHogenboom) [reply tests/TESTING_GUIDE.md]: Fully agree, I've opted to fully remove this file as its a bit redundant: changes in 41d84e9116e0da6e4a1ca996d2b144e93dd8d929

We can still opt to include a 'how to run the tests' note in the docsting of the specific test files, if that feels necessary?

File: tests/TESTING_GUIDE.md (hunk 34 above)
Comment (JHogenboom) [reply tests/TESTING_GUIDE.md]: Same as [above](https://github.com/STRONGAYA/v6-tools-general/pull/8#discussion_r2541114719):
> Fully agree, I've opted to fully remove this file as its a bit redundant: changes in [41d84e9](https://github.com/STRONGAYA/v6-tools-general/commit/41d84e9116e0da6e4a1ca996d2b144e93dd8d929)
> 
> We can still opt to include a 'how to run the tests' note in the docsting of the specific test files, if that feels necessary?

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 7 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/miscellaneous.py:242]: I don't really understand my original comment - sorry, forgot about it. Now that I'm reviewing this code again, I see you use this function often to return an empty dataframe or a value of 0 when data is not available. 
I wonder if that is safe. E.g. if you can't compute the mean, you return that it is 0, but that is not correct (probably). How do you handle these default values down the line to prevent that they skew the analysis results?

File: src/vantage6_strongaya_general/privacy_measures.py (hunk 25 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/privacy_measures.py:1165]: reaction missing?

File: src/vantage6_strongaya_general/general_statistics.py (hunk 32 above)
Comment (bartvanb) [reply src/vantage6_strongaya_general/general_statistics.py:894]: If I look at the code, e.g. [here in descriptive statistics algorithm](https://github.com/STRONGAYA/v6-descriptive-statistics/blob/main/v6-descriptive-statistics/partial.py#L206) I see that it calls the functions within this library, but there are no privacy checks there, nor are they here. Did I miss where they are located?

I don't fully understand the link you shared. That seems like it's testing functions with given parameters, but not this function we're discussing here right?

Comment (bartvanb) [reply src/vantage6_strongaya_general/general_statistics.py:894]: I feel like I'm missing something since you have those tests that expect PrivacyThresholdViolations, but I don't see where those are implemented

File: tests/TESTING_GUIDE.md (hunk 33 above)
Comment (bartvanb) [reply tests/TESTING_GUIDE.md]: Yes, we should say somehwere how the tests can be run. Maybe in the docstring of the tests themselves but preferably in readme / other documentation

File: src/vantage6_strongaya_general/miscellaneous.py (hunk 7 above)
Comment (JHogenboom) [reply src/vantage6_strongaya_general/miscellaneous.py:242]: Okay for use now, to update in a next version as usage can be misleading if not implemented in the algorithm correctly

File: tests/TESTING_GUIDE.md (hunk 33 above)
Comment (JHogenboom) [reply tests/TESTING_GUIDE.md]: Already present in the `README.md` file: https://github.com/STRONGAYA/v6-tools-general#running-tests

**Comment by JHogenboom** [comment]
Keeping this for reference and next major release
